<html>
	<head>
		<title> Table of Contents</title>
		<link rel="stylesheet" href="mev manual.css">
		</style>
		</head>
	<body>
	<div class="titleshot">
	<a href="MEVpic.jpg">
					<img src="MEVpic.jpg">
				</a>
				<br/>
				<span class="title">MEV: MultiExperiment Viewer<br/> Version 4.7.0<br/> May 16, 2011</span>
				</div>
			<hr/>
	<h1> Table of Contents </h1>
	<ul id="toc"><dl>
		<li><a href="GeneralInformation.html"><dt>General Information</dt></a><br/></li>

		<li><a href="MeV Manual.html"><dt>TM4 Software Overview</dt></a><br/></li>

		<li><a href="StartingMEV.html"><dt>Starting MultiExperiment Viewer</dt></a><br/></li>
			<li><a href="StartingMEV.html#Java"><dd>Using Java Webstart<dd></a><br /></li>

		<li><a href="Loading the Data.html"><dt>Loading Expression Data</dt></a><br /></li>
			<li><a href="Loading the Data.html#annotation-feature"><dd>Using the Annotation Feature</dd></a><br /></li>
			<li><a href="Loading the Data.html#IVLD"><dd>Initial View of the Loaded Data, Main Expression Image</dd></a><br /></li>
			<li><a href="Loading the Data.html#RNT"><dd>Result Navigation Tree</dd></a><br /></li>
			<li><a href="Loading the Data.html#HNL"><dd>The History Node and Log</dd></a><br /></li>

			<li><a href="Loading-mev-data.html"><dd>MeV and other TIGR Files</dd></a> <br/></li>
			<li><a href="Loading-affy-data.html"><dd>Affymetrix Files</dd></a> <br/></li>
			<li><a href="Loading-geo-data.html"><dd>Gene Expression Omnibus (GEO) Files</dd></a> <br/></li>
			<li><a href="Loading-magetab-data.html"><dd>MAGE-TAB Files</dd></a> <br/></li>
			<li><a href="Loading-genepix-data.html"><dd>Genepix Files</dd></a> <br/></li>
			<li><a href="Loading-agilent-data.html"><dd>Agilent Files</dd></a> <br/></li>
			<li><a href="Loading-rnaseq-data.html"><dd>RNASeq Data Files</dd></a> <br/></li>


		<li><a href="Adjustingthedata.html"><dt>Adjusting the Data</dd></a><br /></li>
			<li><a href="Adjustingthedata.html#AFO"><dd>Adjustment/Filter Overview</dd></a><br /></li>
			<li><a href="Adjustingthedata.html#DataTrans"><dd>Data Transformations</dd></a><br /></li>
			<li><a href="Adjustingthedata.html#DSS"><dd>Data Source Selection</dd></a><br /></li>

		<li><a href="displayoptions.html"><dt>Display Options</dd></a><br /></li>
			<li><a href="displayoptions.html#SampleAnno"><dd>Sample Annotations</dd></a><br /></li>
			<li><a href="displayoptions.html#SelectGene"><dd>Selecting Gene Annotation</dd></a><br /></li>
			<li><a href="displayoptions.html#Color"><dd>Color Scheme Selection</dd></a><br /></li>
			<li><a href="displayoptions.html#scllimits"><dd>Setting Color Scale Limits</dd></a><br /></li>
			<li><a href="displayoptions.html#EA"><dd>Element Apperance</dd></a><br /></li>

		<li><a href="ViewerDescriptions.html"><dt>Viewer Descriptions</dd></a><br /></li>
			<li><a href="ViewerDescriptions.html#overview"><dd>Overview</dd></a><br /></li>
			<li><a href="ViewerDescriptions.html#ExpressI"><dd>Expression Images</dd></a><br /></li>
			<li><a href="ViewerDescriptions.html#ExpressGraph"><dd>Expression Graphs</dd></a><br /></li>
			<li><a href="ViewerDescriptions.html#CG"><dd>Centroid Graphs</dd></a><br /></li>
			<li><a href="ViewerDescriptions.html#TV"><dd>Table Views</dd></a><br /></li>
			<li><a href="ViewerDescriptions.html#VD"><dd>Venn Diagrams</dd></a><br /></li>
			<li><a href="ViewerDescriptions.html#expressChart"><dd>Expression Charts</dd></a><br /></li>
			<li><a href="ViewerDescriptions.html#CVA"><dd>Common Viewer Activities</dd></a><br /></li>

		<li><a href="WorkingwithClusters.html"><dt>Working with Clusters</dd></a><br /></li>
			<li><a href="WorkingwithClusters.html#SCUCM"><dd>Storing Clusters and Using the Cluster Manager</dd></a><br /></li>

		<li><a href="UM.html"><dt>Utilities Menu</dd></a><br /></li>
			<li><a href="UM.html#Search"><dd>Search Utility</dd></a><br /></li>
			<li><a href="UM.html#IGSL"><dd>Import Gene or Sample List</dd></a><br /></li>
			<li><a href="UM.html#ASA"><dd>Append Sample Annotation</dd></a><br /></li>
			<li><a href="UM.html#AGA"><dd>Append Gene Annotation</dd></a><br /></li>

		<li><a href="createout.html"><dt>Creating Output</dd></a><br /></li>
			<li><a href="createout.html#SaveAny"><dd>Saving the Analysis</dd></a><br /></li>
			<li><a href="createout.html#SaveMate"><dd>Saving the Expression Matrix</dd></a><br /></li>
			<li><a href="createout.html#SaveView"><dd>Saving Viewer Images</dd></a><br /></li>
			<li><a href="createout.html#SaveClust"><dd>Saving Cluster Data</dd></a><br /></li>

		<li><a href="gag.html"><dt>The Gaggle Implementation</dd></a><br /></li>
			<li><a href="gag.html#IntroGag"><dd>Introduction to the Gaggle</dd></a><br /></li>
			<li><a href="gag.html#GagMenu"><dd>The Gaggle Menu</dd></a><br /></li>
			<li><a href="gag.html#Gagmev"><dd>Using the Gaggle with MeV</dd></a><br /></li>

		<li><a href="Mod.html"><dt>Modules</dd></a><br /></li>
			<li><a href="Mod.html#DCGP"><dd>Description Conventions and General Pointers</dd></a><br /></li>

						<h4>Clustering</h4>
		<li><a href="HCL.html"><dt>HCL: Hierarchical Clustering</dt></a><br /></li>
			<li><a href="HCL.html#hcl_params"><dd>HCL Parameters</dd></a><br /></li>
			<li><a href="HCL.html#adjust"><dd>Adjusting the Tree Configuration and Viewing Clusters</dd></a><br /></li>

		<li><a href="TEASE.html"><dt>TEASE: Tree-Ease</dd></a><br /></li>
			<li><a href="TEASE.html#mode"><dd>Mode selection</dd></a><br /></li>
			<li><a href="TEASE.html#para"><dd>TEASE Parameters</dd></a><br /></li>
			<li><a href="TEASE.html#Navi"><dd>Navigating the Hierarchical Tree</dd></a><br /></li>
			<li><a href="TEASE.html#Adjust1"><dd>Adjusting Tree Configuration and Viewing Clusters</dd></a><br /></li>

		<li><a href="ST.html"><dt>ST: Support Trees</dd></a><br /></li>
			<li><a href="ST.html#para"><dd>Parameters</dd></a><br /></li>

		<li><a href="SOTA.html"><dt> SOTA: Self Organizing Tree Algorithm </dd></a><br /></li>
			<li><a href="SOTA.html#para"><dd>Parameters and Basic Terminology</dd></a><br /></li>

		<li><a href="KMC.html"><dt> KMC: K-Means/K-Medians Clustering </dd></a><br /></li>
			<li><a href="KMC.html#para"><dd> Parameters </dd></a><br /></li>

		<li><a href="kms.html"><dt>KMS: K-Means/K-Medians Support  </dd></a><br /></li>
			<li><a href="kms.html#para"><dd>Parameters</dd></a><br /></li>

		<li><a href="CAST.html"><dt>CAST: Clustering Affinity Search Technique </dd></a><br /></li>
			<li><a href="CAST.html#para"><dd>Parameters </dd></a><br /></li>

		<li><a href="FOM.html"><dt>FOM: FOM-Figures of Merit</dd></a><br /></li>


		<li><a href="QTC.html"><dt>QTC: QT Cluster </dd></a><br /></li>
			<li><a href="QTC.html#para"><dd>Parameters</dd></a><br /></li>

		<li><a href="SOM.html"><dt>SOM: Self Organizing Maps</dd></a><br /></li>
			<li><a href="SOM.html#basic"><dd>Basic Terminology</dd></a><br /></li>
			<li><a href="SOM.html#para"><dd>Parameters</dd></a><br /></li>

		<li><a href="NMF.html"><dt>NMF: Non-negative Matrix Factorization</dd></a><br /></li>
			<li><a href="NMF.html#para"><dd>Parameters</dd></a><br /></li>

							<h4>Statistics</h4>
		<li><a href="ptm.html"><dt>PTM: Template Matching </dd></a><br /></li>
			<li><a href="ptm.html#para"><dd>Parameters </dd></a><br /></li>

		<li><a href="test.html"><dt>TTEST: T-tests </dd></a><br /></li>
			<li><a href="test.html#para"><dd>Parameters</dd></a><br /></li>

		<li><a href="sam.html"><dt>SAM: Significance Analysis of Microarrays</dd></a><br /></li>
			<li><a href="sam.html#para"><dd>Parameters</dd></a><br /></li>

		<li><a href="anova.html"><dt>ANOVA: Analysis of Variance</dd></a><br /></li>
			<li><a href="anova.html#para"><dd>Parameters</dd></a><br /></li>

		<li><a href="anova2.html"><dt>TFA: Two-factor ANOVA</dd></a><br /></li>

		<li><a href="nonpar.html"><dt>NonpaR: Nonparametric Statistical Tests </dd></a><br /></li>
			<li><a href="nonpar.html#des"><dd>Brief Test Descriptions</dd></a><br /></li>
			<li><a href="nonpar.html#run"><dd>Running NonpaR Tests</dd></a><br /></li>
			<li><a href="nonpar.html#mack"><dd>Mack-Skillings (MS) Test Dialogs</dd></a><br /></li>
			<li><a href="nonpar.html#fish"><dd>Fisher Exact Test</dd></a><br /></li>

		<li><a href="betr.html"><dt>BETR: Bayesian Estimation of Temporal Regulation </dd></a><br /></li>
			<li><a href="betr.html#para"><dd>Parameters</dd></a><br /></li>

		<li><a href="limma.html"><dt>LIMMA: Linear Models for Microarray Data</dd></a><br /></li>

		<li><a href="goseq.html"><dt>GOSeq: Gene Ontology Analysis for RNASeq</dd></a><br /></li>

		<li><a href="surv.html"><dt>SURV: Survival Analysis</dd></a><br /></li>
			<li><a href="surv.html#run"><dd>Running the Module </dd></a><br /></li>

		<li><a href="gcova.html"><dt>GLOBANC: Global Analysis of Covariance</dd></a><br /></li>

		<li><a href="rp.html"><dt>RP: Rank Products</dd></a><br /></li>
			<li><a href="rp.html#run"><dd>Running the Module</dd></a><br /></li>
			<li><a href="rp.html#para"><dd>Parameters</dd></a><br /></li>

		<li><a href="minet.html"><dt>MINET: Mutual Information Network</dd></a><br /></li>

		<li><a href="edger.html"><dt>EdgeR: Empirical Analysis of RNASeq Data in R</dd></a><br /></li>

		<li><a href="deseq.html"><dt>DeSEQ: Analysis of RNASeq Data based on Negative Binomial Distribution</dd></a><br /></li>

		<li><a href="degseq.html"><dt>DEGSEQ: Analysis of RNASeq Data based on MA Plot</dd></a><br /></li>

							<h4>Classification</h4>
		<li><a href="svm.html"><dt>SVM: Support Vector Machines</dd></a><br /></li>
			<li><a href="svm.html#overview"><dd>SVM Dialog Overview</dd></a><br /></li>
			<li><a href="svm.html#para"><dd>Process Initialization Parameter Information</dd></a><br /></li>
			<li><a href="svm.html#out"><dd>SVM Output</dd></a><br /></li>

		<li><a href="usc.html"><dt>USC: Uncorrelated Shrunken Centroids</dd></a><br /></li>
			<li><a href="usc.html#para"><dd>Advanced Parameters</dd></a><br /></li>

		<li><a href="knn.html"><dt>KNNC: K-Nearest-Neighbor Classification</dd></a><br /></li>
			<li><a href="knn.html#para"><dd>KNN Classification Parameter Information</dd></a><br /></li>

		<li><a href="dam.html"><dt>DAM: Discriminant Analysis Module</dd></a><br /></li>
			<li><a href="dam.html#para"><dd>Parameters</dd></a><br /></li>

			<h4>Data Reduction</h4>

		<li><a href="RN.html"><dt> RN: Relevance Networks </dd></a><br /></li>
			<li><a href="RN.html#para"><dd> Parameters </dd></a><br /></li>

		<li><a href="pca.html"><dt> PCA: Principal Components Analysis  </dd></a><br /></li>

		<li><a href="coa.html"><dt> COA: Correspondence Analysis </dd></a><br /></li>

		<li><a href="trn.html"><dt> TRN:Expression Terrain Maps </dd></a><br /></li>
			<li><a href="trn.html#tn"><dd> Terrain Navigation </dd></a><br /></li>
			<li><a href="trn.html#tvm"><dd> The Terrain Viewer Menu </dd></a><br /></li>
			<li><a href="trn.html#cluster"><dd>Cluster Selection and Related Operations</dd></a><br /></li>

		<h4>Meta Analysis</h4>
		<li><a href="gsea.html"><dt> GSEA:Gene Set Enrichment Analysis </dd></a><br /></li>

		<li><a href="attr.html"><dt> ATTR:Attract Package  </dd></a><br /></li>

		<li><a href="ease.html"><dt> EASE: Expression Analysis Systematic Explorer  </dd></a><br /></li>
			<li><a href="ease.html#para"><dd> Parameters  </dd></a><br /></li>
			<li><a href="ease.html#apara"><dd> Advanced Parameters Dialog  </dd></a><br /></li>
			<li><a href="ease.html#statspara"><dd> Statistical Parameters Dialog </dd></a><br /></li>

								<h4>Visualization</h4>
		<li><a href="lem.html"><dt>LEM: Linear Expression Maps  </dd></a><br /></li>

		<li><a href="gdm.html"><dt> GDM: Gene Distance Matrix  </dd></a><br /></li>
			<li><a href="gdm.html"><dd> Matrix Viewer Basics  </dd></a><br /></li>

		<li><a href="bn.html"><dt> BN: Bayesian Networks Analysis </dd></a><br /></li>

		<li><a href="lm.html"><dt> LM:Literature Mining Analysis  </dd></a><br /></li>
		<br/>

								<h4>Miscellaneous</h4>
		<li><a href="gsh.html"><dt> GSH: Gene Shaving  </dd></a><br /></li>
		<br/>

		<li><a href="script.html"><dt> Scripting </dd></a><br /></li>
			<li><a href="script-new-script.html"><dd> The Script  </dd></a><br /></li>
			<li><a href="script-diversity-filter.html"><dd> Diversity Variance Filter  </dd></a><br /></li>
			<li><a href="script-centroid-variance-filter.html"><dd> Centroid Variance Filter  </dd></a><br /></li>
		<br/>
		<li><a href="compare.html"><dt> Comparative Genomic Hybridization Viewer </dd></a><br /></li>
			<li><a href="compare.html#exp"><dd> Loading Experiments </dd></a><br /></li>
			<li><a href="compare.html#view"><dd>CGH Analyzer Viewers </dd></a><br /></li>
			<li><a href="compare.html#graph"><dd> The CGH Position Graph</dd></a><br /></li>
			<li><a href="compare.html#brow"><dd> The CGH Browser</dd></a><br /></li>
			<li><a href="compare.html#any"><dd> CGH Analysis</dd></a><br /></li>
			<li><a href="compare.html#hl"><dd> Higher Level Analysis</dd></a><br /></li>
								<h4>Appendix</h4>
		<li><a href="app.html"><dt> Appendix: File Format Descriptions </dd></a><br /></li>
			<li><a href="app.html#tav"><dd> TAV Files </dd></a><br /></li>
			<li><a href="app.html#tdms"><dd> Tab Delimited, Multiple Sample Files (TDMS files)</dd></a><br /></li>
			<li><a href="app.html#gene"><dd> GenePix Files </dd></a><br /></li>
			<li><a href="app.html#mev"><dd> MEV Files</dd></a><br /></li>
			<li><a href="app.html#anno"><dd> Annotation Files (.ann) </dd></a><br /></li>
			<li><a href="app.html#bio"><dd>Bioconductor (MAS5) Files </dd></a><br /></li>
			<li><a href="app.html#affy"><dd>Affymetrix GCOS (Pivot Data) File </dd></a><br /></li>
			<li><a href="app.html#geo"><dd>GEO SOFT Affymetrix File Format </dd></a><br /></li>
			<li><a href="app.html#geos"><dd>GEO SOFT two channel file format </dd></a><br /></li>
			<li><a href="app.html#dchip"><dd>dChip or DFCI core file format </dd></a><br /></li>
			<li><a href="app.html#afss"><dd>Assignment File Saving System </dd></a><br /></li>
			<li><a href="app.html#123"><dd>Special Cases </dd></a><br /></li>

		<li><a href="app2.html"><dt> Appendix: Preferences Files </dd></a><br /></li>

		<li><a href="app3.html"><dt> Appendix: Distance Metrics</dd></a><br /></li>

		<li><a href="app4.html"><dt> Appendix: MeV Script DTD</dd></a><br /></li>
			<li><a href="app4.html#dtd"><dd>DTD UML Schema </dd></a><br /></li>

		<li><a href="app5.html"><dt> Appendix: MeV R Integration</dd></a><br /></li>

		<li><a href="app6.html"><dt>Appendix: Bayesian Network & Literature Mining</dd></a><br /></li>

		<li><a href="app7.html"><dt>Appendix: MeV Dependencies</dd></a><br /></li>
		<br/>

		<li><a href="license.html"><dt>License</dd></a><br /></li>

		<li><a href="contri.html"><dt>Contributors</dd></a><br /></li>

		<li><a href="oome.html"><dt>Out of Memory Errors</dd></a><br /></li>
		<li><a href="hcl_newick_output_parameters.html"><dt>Newick tree output</dd></a><br /></li>
		<li><a href="hcl_nexus_output_parameters.html"><dt>Nexus tree output</dd></a><br /></li>



		</dl></ul>
		</body>
		</html>