<html>
	<body>
		<h1>
			<b>The TM4 Software Development Team  expresses sincere thanks for 
				the contributions of those listed below.
		</b></h1>
			<table border="0" >
				<tr align="left">
					<th>
					<b>Contributor</b> 
					</th>
                    <th>
                    	<b>Affiliation</b>
                    </th>
                    <th>
                    	<b>Contribution</b>
                    </th>
                </tr>
				
                <tr valign="top">
                	<td>Nirmal Bhagabati, John Braisted, Vasily Sharov, Wei Liang, Chun-Hua Wan, and Alexander I. Saeed
                	</td>
                	<td>The Institute for Genomic Research / Pathogen Functional Genomics Resource Center
                	</td>
                	<td>System design, module development and implementation, project coordination, documentation, optimization, usability assurance, user support, and Linear Expression Map (LEM) module development.
                	</td>
                </tr>	
				
                <tr valign="top">
                	<td>Eleanor Howe, Sarita Nair, John Quackenbush, Raktim Sinha, Dan Schlauch, Weidong Wang
                	</td>
                	<td>Dana-Farber Cancer Institute
                	</td>
                	<td>Analysis-saving rewrite, v4.0 release coordination, graphics, documentation, CGH data analysis package, additional data loaders, other miscellaneous enhancements.
                	</td>
                </tr>
                <tr valign="top">
                	<td>Vu Chu, Annie Liu, Roger Bumgarner
                	</td>
                	<td>University of Washington
                	</td>
                	<td>Agilent data loader, USC, RAMA, Bridge, TEASE
                	</td>
                </tr>
                <tr valign="top">
                	<td>Raktim Sinha
                	</td>
                	<td>Dana Farber Cancer Institute
                	</td>
                	<td>Added aCGH type data loaders, aCGH visualizing modules and aCGH analysis
						algorithms functionality to MeV based on a published open source tool &amp;
						article(Adam A. Margolin et al, Bioinformatics 2005)
                	</td>
                </tr>

										

				<!--
				//TODO
				Raktim Sinha
Raktim_Sinha@dfci.harvard.edu
Bioinformatics Analyst, DFCI
Added aCGH type data loaders, aCGH visualizing modules and aCGH analysis
algorithms functionality to MeV based on a published open source tool &
article(Adam A. Margolin et al, Bioinformatics 2005)

UW Credits -
Roger Bumgarner is the PI of the UW group's efforts in MeV
Vu Chu coded: The Agilent data loader, USC, RAMA, and BRIDGE
USC was based on code from Ka Yee Yeung
RAMA and BRIDGE make use of code developed by Raphael Gottardo and a
collaboration between Yeung, Bumgarner, Gottardo and Adrian Raftery.
Annie Liu did the coding for TEASE


				latest from TIGR.  Need to figure out how to reconcile this with above entry
				Nirmal Bhagabati, John Braisted, Vasily Sharov, Wei Liang, Chun-Hua Wan, and Alexander I. Saeed	The Institute for Genomic Research /
Pathogen Functional Genomics Resource Center	System design, module development and implementation, project coordination, documentation, optimization, usability assurance, user support, and Linear Expression Map (LEM) module development
				-->
                <tr valign="top">
                	<td>Alexander Sturn, Zlatko Trajanoski
                	</td>
                	<td>Institute of Biomedical Engineering, Graz University of Technology
                	</td>
                	<td>Initial system architecture and viewer design, initial module development (HCL, KMC, SOM, PCA, SVM)
                	</td>
                </tr>
                <tr valign="top">
                	<td>Mark Snuffin, Aleksey Rezantsev, Dennis Popov, Alex Ryltsov, Edward Kostukovich, Igor Borisovsky
                	</td>
                	<td>DataNaut, Inc.
                	</td>
                	<td>System architecture, module architecture, parallel processing system, primary module development (RN, TRN)
                	</td>
                </tr>
                <tr valign="top">
                	<td>Stu Golub, Zaigang Liu, Jane Ruan, Minhas Siddiqui
                	</td>
                	<td>Syntek Systems Corporation, Inc.
                	</td>
                	<td>Primary module development (CAST, FOM, GSH, GDM)
                	</td>
                </tr>
                <tr valign="top">
                	<td>Wei Liang
                	</td>
                	<td>The Institute for Genomic Research
                	</td>
                	<td>Normalization algorithm development, TM4 website maintenance
                	</td>
                </tr>
                <tr valign="top">
                	<td>Jerry Li, Vasily Sharov, Joe White, Mathangi Thiagarajan, Tracey Currier, Eleanor Howe
                	</td>
                	<td>The Institute for Genomic Research
                	</td>
                	<td>Additional development, software testing, documentation, support
                	</td>
                </tr>
                <tr valign="top">
                	<td>Patrick Cahan, Tim McCaffrey
                	</td>
                	<td>The George Washington University
                	</td>
                	<td>Affymetrix data loader and filters, Affymetrix loader rework, gene submission capability for 
                		the LOLA cluster repository
                	</td>
                </tr>
                <tr valign="top">
                	<td>Todd Peterson
                	</td>
                	<td>National Center for Genome Resources
                	</td>
                	<td>GeneX-Lite to MeV connectivity
                	</td>
                </tr>
                <tr valign="top">
                	<td>Luke Somers
                	</td>
                	<td>Fox Chase Cancer Center
                	</td>
                	<td>QTC algorithm optimization and bug fix
                	</td>
                </tr>
                <tr valign="top">
                	<td>Jim Johnson, Ernest Retzel
                	</td>
                	<td>Center for Computational Genomics and Bioinformatics, University of Minnesota
                	</td>
                	<td>Java WebStart configuration, scripts and consultation to support using ANT technologies for MeV development
                	</td>
				</tr>
                <tr valign="top">
                	<td>Glynn Dennis, Douglas Hosack, Richard Lempicki, Wei Gao
                	</td>
                	<td>National Institute of Allergy and Infectious Disease, NIH, Laboratory of Immunopathogenesis and Bioinformatics.
                	</td>
                	<td>EASEOpenSource Java package to support EASE development and helpful consultation during EASE integration and development.
                	</td>
                </tr>
                <tr valign="top">
                	<td>Eric Albert
                	</td>
                	<td>Independent Development
                	</td>
                	<td>Implementation of BrowserLauncher.java, utility to launch default web browser from within MeV
                	</td>
                </tr>
                <tr valign="top">
                	<td>Sally Gaddis
                	</td>
                	<td>University of Texas, MD Anderson Cancer Center
                	</td>
                	<td>Report and basic testing of Java3D support for mev 3D viewers on the Mac platform.
                	</td>
                </tr>
                <tr valign="top">
                	<td>Stephen C. Harris
                	</td>
                	<td>FDA, National Center for Toxicological Research
                	</td>
                	<td>T-test bug identification, and suggestion for improvement of adjusted Bonferroni correction.
                	</td>
                </tr>
			</table>
	</body>
</html>