<html>
	<head>
		<title> ATTR: Attract Package   </title>
		<link rel="stylesheet" href="mev manual.css">
		</head>
		</style>
		<body>
<div class="navigation-bar">
	<a  class="prev" href="gsea.html">Previous: GSEA-Gene Set Enrichment Analysis </a> | 
	<a  class="toc" href="TableofContents.html">Table of Contents</a> | 
	<a  class="next" href="ease.html">Next: EASE-Expression Analysis Systematic Explorer </a>
</div>
	<h3> Attract: Attract Package  </h3>
		<p>(Mar JC, Wells CA, Quackenbush J. Defining an informativeness metric for clustering gene expression data. <i>Bioinformatics.</i> 2011 Apr 15;27(8):1094-100.J. PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/21330289">21330289</a>.)</p>
<h5>What is Attract?</h5>
	<p>The Attract algorithm identifies the core gene expression modules that are differentially activated between cell types or different sample groups, and elucidates the set of expression profiles which describe the range of transcriptional behavior within each module. </h5>
	<p>The algorithm assists in understanding expression changes that are specific to cell types or cell fate transitions. A mammalian organism is made up of over 200 types of specialized cells. Each cell type carries out a specific task integral to maintaining homeostasis of the organism. Cell types can vary by morphology, structure, lifespan, functional ability and much more. Different cell or tissue types acquire their diversity by driving differentially coordinated expression patterns through interacting gene networks. The attractor hypothesis proposed by Stuart Kauffman describes how cell fate transitions between cell types occur through coordinated changes in genome-wide gene expression. </p>
	<p>The Attract algorithm starts by using a GSEA step to identify the set of core pathway modules driving the cell types-specific expression changes. For each of these modules, the algorithm then constructs synexpression groups, which consist of genes with correlated expression profiles. The algorithm is also able to extrapolate out to identify genes with similar expression profiles which may not be currently annotated in current pathway databases like KEGG. </p>
	<h5>When should Attract be used</h5>

	<p>The Attract algorithm can be used for any expression data set where replicate samples (at least three) have been collected for various cell types or experimental groups such as a panel of drug treatments or a time course experiment. The algorithm is not well-suited for data generated under an experimental design that has only one experimental or covariate group, e.g. samples from 100 identical cell lines with no perturbations.</p>
<h5>Brief Description of the algorithm</h5>
<p>The Attract algorithm can be roughly divided in to four steps:</p>
<dl><ol>
<dt>1. Find core attractor state pathway modules in your data using GSEA.</dt>
<dt>2. Remove flat or uninformative genes (i.e. genes that do not show expression changes across the different factor levels).  </dt>
<dt>3. Find Synexpression  groups. (A synexpression group contains genes that share similar expression profiles across factor levels)</dt>
	<dd>--	Use the informativeness metric to compute clusters (synexpression groups) in each geneset. </dd>
<dt>4. For every synexpression group in a geneset,  find genes in the expression data which have correlated expression profiles.</dt>
</ol></dl> 

	<div class="screenshots leftsidecorrection">
		<a href="attr1.png">
			<img src="attr1.png">
		</a>
		<br/>
		<span class="caption">Figure 1: The initialization dialog</span>
	</div>
	<h5>Running the Attract Module</h5>

	<p>Choose the type of analysis you wish to run. Attract supports two-class analysis for comparison of two groups of samples, or multi-class analysis, for comparison of three or more groups. 
	</p>
	<p>
	Choose an alpha cutoff level. This number is a false discovery rate used to remove flat or uninformative genes. 
	</p>
	<p>
	From the drop-down menu, select the name of the microarray platform your data was collected from. This step is very important. 
	</p>
	<p>
	Click <i>Continue</i> to display the Cluster Selector.
	</p>

	<p>If you have previously created clusters of samples using the Cluster Manager or another MeV module, they will be displayed here, as in Figure 2. Use the drop-down menus on the right-hand side of the Cluster Selector to assign your sample groups to analysis groups for Attract. You can assign more than one sample group to an analysis group. 	You can also use the <span class="widget2">“Button Selection”</span> tab shown in Figure 3 to achieve the same. Group1 and Group2 symbolize B-ALL and T-ALL. You can save these grouping using the <span class="widget2">“Save settings”</span> button. To load saved groupings, use the <span class="widget2">“Load settings”</span> button. <span class="widget2">Reset</span> button will clear all your choices. 
	Click the <span class="widget2">Ok</span> button to run Attract. 
	<div class="screenshots">
		<a href="attr2.png">
			<img src="attr2.png">
		</a>
		<br/>
		<span class="caption">Figure 2: The Attract Dialog with cluster selector</span>
	</div>
	
	<h5>Attract Results</h5>
	
	The results of the Attract module are found in the Result Tree, on the left-hand side of the Multiple Array Viewer window. 
	The <widget2>Results Table</widget2> contains a list ....
	<div class="screenshots">
		<a href="attr3.png">
			<img src="attr3.png">
		</a>
		<br/>
		<span class="caption">Figure 3: Attract Results</span>
	</div>
	
	<div class="navigation-bar">
		<a class="prev" href="gsea.html">Previous: GSEA-Gene Set Enrichment Analysis </a> | 
		<a  class="toc" href="TableofContents.html">Table of Contents</a> | 
		<a  class="next" href="ease.html">Next: EASE-Expression Analysis Systematic Explorer </a>
	</div>
	
</body>
</html>		
		
		
		