<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "DTD/xhtml1-transitional.dtd">
<html>
	<head>
		<title>MeV File Loader - RNASeq File Format</title>
		<link rel="stylesheet" href="mev_manual.css">
	</head>
	<body>
   		<div class="navigation-bar">
			<a href="Loading-agilent-data.html">Previous: Loading Agilent Data</a> | 
			<a class="toc" href="TableofContents.html">Table of Contents</a> | 
			<a class="next" href="Adjustingthedata.html">Next: Adjusting the Data</a>
		</div>
        <h3>RNASeq File Loader - Overview</h3>
        <p>
            Count and RPKM  File Format is a <b>tab-delimited</b> file format for loading HTS(High Throughput Sequencing Data)
            into MeV for analysis. We are currently providing RNASeq analysis algorithms only for these data sets.
        </p>
        <p>
            MeV being a desktop software cannot provide functionalities like base calling and sequence alignment etc, which are computationaly intensive
            processes and should be be done on hign end server clusters.  
        </p>
        <p>
            The entry point for HTS data into MeV should be in a <b>summarized</b> form. Which means, the raw sequence data has been base-called
            and aligned and tag counts have been assembled, summarized and mapped to a reference genome at the transcript/gene level.
        </p>
        <p>
            This loader supports both Counts and RPKM/FPKM and a combinations of both as described in the Data Type section below. Currently
            we have support for Human and Mouse data only and it is provided as annotaiton from RefSeq or ENSMBL.
        </p>
        <p>
            Both Count &amp; Expression (RPKM/FPKM) are maintained for this kind of loader. The user has the option of loading either or both
            kinds of info for there data set. If either Count or RPKM is left out, it is calculated based on the publication descibed in the
            section <b>Count to RPKM and vice versa</b>.
        </p>
        
        <h4>Dialog Selections</h4>
        <ul>
			<li>
                <b>Upload User Annotation []</b>: 
				<p>Currently we porvide annotation support for Human ans Mouse data only and would add support
				   for more organisms in the future. However the user can provide annotaiton for any organism they want as they follow the
				   annotation file format specifications. By selecting this option the user agrees to upload a annotation file which matches
				   the example file provided in the <b>data/rnaseq/ref_gene_h19_sample_anno.txt</b>. The matching data file is also in the same
				   location, <b>data/rnaseq/TagSeqExample.txt</b>. The table below describes the header information for an annotation file:
                </p>
				<p> 
                    The <b>15 annotaton columns</b> are described here. They columns should appear in file in the <b>same order</b> as described here:
                    <table border='1'>
                        <thead>
                            <tr>
                                <td bgcolor='cream'><b>##</b></td>
                                <td bgcolor='#FFA500'><b>Field Name</b></td>
                                <td bgcolor='#ADD8E6'><b>Description</b></td>
                            </tr>
                        </thead>
                        <tr>
                            <td><b>1</b></td>
                            <td><b>PROBE_ID</b></td>
                            <td><b>Required.</b> A unique ID the identifies each row in the data file.It is treated like the probe_id of microarray data.
								To be associated with a the data being loaded the <b>nearest_ref_id</b> filed in the data file should correspond to this
								field in the annotation file. This is generally expected to be a RefSeq or Ensemble ID but user can use anything as long
								they correspond in the data and the annotation file.
							</td>
                        </tr>
                        <tr>
                            <td><b>2</b></td>
                            <td><b>CHR</b></td>
                            <td><b>Required</b> but an empty value is ok. Stands for chromosome and when provide the format should be: <b>chr1</b></td>
                        </tr>
                        <tr>
                            <td><b>3</b></td>
                            <td><b>STRAND</b></td>
                            <td><b>Required.</b> When not avaialable for a row a '-' or '' can be used.</td>
                        </tr>
                        <tr>
                            <td><b>4</b></td>
                            <td><b>TX_START</b></td>
                            <td><b>Required.</b> but an empty value is ok. When provided should indicate the BP position. </td>
                        </tr>
                        <tr>
                            <td><b>5</b></td>
                            <td><b>TX_END</b></td>
                            <td><b>Required</b>, but an empty value is ok. When provided should indicate the BP position.</td>
                        </tr>
						<tr>
                            <td><b>6</b></td>
                            <td><b>CDS_START</b></td>
                            <td><b>Required</b>, but an empty value is ok. Stands for coding sequence start position and when provided should 
								indicate the BP position.
							</td>
                        </tr>
						<tr>
                            <td><b>7</b></td>
                            <td><b>CDS_END</b></td>
                            <td><b>Required</b>, but an empty value is ok. Stands for coding sequence end position and When provided should 
								indicate the BP position.
							</td>
                        </tr>
						<tr>
                            <td><b>8</b></td>
                            <td><b>exonCount</b></td>
                            <td><b>Required</b>, but an empty value is ok. Stands for count of exons in the coding sequence region, 
								and when provided should be an integer value. <b>Currently unused.</b>
							</td>
                        </tr>
						<tr>
                            <td><b>9</b></td>
                            <td><b>exonStarts</b></td>
                            <td><b>Required</b>, but an empty value is ok. Stands for start BP postion of each exon, and when provided should be 
								integer value(s) separated by commas and number of entries should match the exonCount. The format should be 
								<b>29284557,29293459,</b>. <b>Currently unused.</b>
							</td>
                        </tr>
						<tr>
                            <td><b>10</b></td>
                            <td><b>exonEnds</b></td>
                            <td><b>Required</b>, but an empty value is ok. Stands for end BP postion of each exon, and when provided should be 
								integer value(s) separated by commas and number of entries should match the exonCount. The format should be 
								<b>29284557,29293459,</b>.<b>Currently unused.</b></td>
                        </tr>
						<tr>
                            <td><b>11</b></td>
                            <td><b>GENE_SYMBOL</b></td>
                            <td><b>Required</b>, but an empty value is ok. Stands for gene Symbol.</td>
                        </tr>
						<tr>
                            <td><b>12</b></td>
                            <td><b>GENE_TITLE</b></td>
                            <td><b>Required</b>, but an empty value is ok. Stands for gene tile/description.</td>
                        </tr>
						<tr>
                            <td><b>13</b></td>
                            <td><b>REFSEQ_ACC</b></td>
                            <td><b>Required</b>, but an empty value is ok. Stands for accession number like RefSeq Id or Ensembl Id.</td>
                        </tr>
						<tr>
                            <td><b>14</b></td>
                            <td><b>PROTEIN_ACC</b></td>
                            <td><b>Required</b>,but an empty value is ok. Stands for protein accession number based on RefSeq Id or Ensembl Id.</td>
                        </tr>
						<tr>
                            <td><b>15</b></td>
                            <td><b>ENTREZ_ID</b></td>
                            <td><b>Required</b>, but an empty value is ok. Stands for Gene Id.</td>
                        </tr>
                    </table>
                </p>
				<p>
					<b>Anootation File Required Characteristics</b>
				</p>
                <ol>
                    <li><b>Tab</b> Delimited</li>
                    <li><b>PRODE_ID</b> in annotation file is expected to match <b>nearest_ref_id</b> in data file.</li>
					<li>All columns are required in the order they are specified but blank/empty values are accepted in most</li>
					<li>Some columns are unused at this time. Empty columns are ok for those. The table above lists unused columns.</li>
					<li>Data rows with unmatched annoation row will still be loaded</li>
					<li>If multiple annotaion rows matches one data row, the first one is used.</li>
                </ol>
				<p>
				<b> Annotation File E.g.:</b>
				<table border='1'>
                        <thead>
                            <tr>
                                <td bgcolor='cream'><b>PROBE_ID</b></td>
                                <td bgcolor='#ADD8E6'><b>CHR</b></td>
                                <td bgcolor='#ADD8E6'><b>STRAND</b></td>
								<td bgcolor='#ADD8E6'><b>TX_START</b></td>
                                <td bgcolor='#ADD8E6'><b>TX_END</b></td>
                                <td bgcolor='#ADD8E6'><b>CDS_START</b></td>
								<td bgcolor='#ADD8E6'><b>CDS_END</b></td>
                                <td bgcolor='#FFA500'><b>exonCount</b></td>
                                <td bgcolor='#FFA500'><b>exonStarts</b></td>
								<td bgcolor='FFA500'><b>exonEnds</b></td>
                                <td bgcolor='#ADD8E6'><b>GENE_SYMBOL</b></td>
                                <td bgcolor='#ADD8E6'><b>GENE_TITLE</b></td>
								<td bgcolor='#ADD8E6'><b>REFSEQ_ACC</b></td>
                                <td bgcolor='#ADD8E6'><b>PROTEIN_ACC</b></td>
                                <td bgcolor='#ADD8E6'><b>ENTREZ_ID</b></td>
                            </tr>
                        </thead>
						<tr>		
							<td>NR_024227</td>
							<td>chr19</td>
							<td>-</td>
							<td>50595745</td>
							<td>50595866</td>
							<td> </td>
							<td> </td>
							<td> </td>
							<td> </td>
							<td> </td>
							<td>NAR-A6S</td>
							<td>Some gene</td>
							<td>NR_024227</td>
							<td></td>
							<td>100169957</td>
							
						</tr>
						<tr>
							<td>NM_024328</td>
							<td>chr14</td>
							<td>+</td>
							<td>24025197</td>
							<td>24028786</td>
							<td>24025966</td>
							<td>24028049</td>
							<td>2</td>
							<td>24025197,24027903,</td>
							<td>24026513,24028786,</td>
							<td>THTPA</td>
							<td>thiamine-triphosphatase</td>
							<td>NM_024328</td>
							<td>NP_077304</td>
							<td>79178</td>
						</tr>
				</table>
				</p>
            </li>
            <li>
                <b>Data Type</b>: 
                <p>
                    We allow the following type of data formats and combinations to be loaded. Each type specifies a file format and
                    what kind of data the program is loading. All <b>4 formats</b> starts with the <b>same 5</b> annotaton columns. 
                    The file formats differ only in the data column(s) and data types; integer or float.
                </p>
                <p> 
                    The <b>5 annotaton columns</b> are described here. They columns should appear in file in the <b>same order</b> as described here:
                    <table border='1'>
                        <thead>
                            <tr>
                                <td bgcolor='cream'><b>##</b></td>
                                <td bgcolor='#FFA500'><b>Field Name</b></td>
                                <td bgcolor='#ADD8E6'><b>Description</b></td>
                            </tr>
                        </thead>
                        <tr>
                            <td><b>1</b></td>
                            <td><b>tracking_id</b></td>
                            <td><b>Required.</b> A unique ID the identifies each row in the data file.</td>
                        </tr>
                        <tr>
                            <td><b>2</b></td>
                            <td><b>locus</b></td>
                            <td><b>Required</b> for each gene/row in the data file. The format should be: <b>chr1:7838183-7838231</b></td>
                        </tr>
                        <tr>
                            <td><b>3</b></td>
                            <td><b>nearest_ref_id</b></td>
                            <td><b>Required.</b> When not avaialable for a row a '-' can be used. This field is used as a key accessioin to
                            link into ReSeq or ENSMBL Dbs for known genomic regions.</td>
                        </tr>
                        <tr>
                            <td><b>4</b></td>
                            <td><b>class_code</b></td>
                            <td><b>Required.</b> , but an empty value is ok. This column can be used for any kind of notes/status about the gene.</td>
                        </tr>
                        <tr>
                            <td><b>5</b></td>
                            <td><b>transcript_length</b></td>
                            <td><b>Required</b>, but an empty value is ok. When not provided, the diff between the start and end BP in
                            the <b>locus</b> is used as the length.</td>
                        </tr>
                    </table>
                </p>
                <ol>
                    <li>
                        <b>RPKM</b>
                        <p>
                            Normalized Expression data in Reads per Kilobase per million unit (RPKM). 
                            The corresponding Count info is calculated by the progema. However, the
                            user is required to <b>provide a 'Library Size file'</b> which should
                            list the size of each sample library. An example:
                        </p>
                        <table border='1'>
                            <thead>
                                <tr>
                                    <td bgcolor='#C0C0C0'><b>tracking_id</b></td>
                                    <td bgcolor='#C0C0C0'><b>locus</b></td>
                                    <td bgcolor='#C0C0C0'><b>nearest_ref_id</b></td>
                                    <td bgcolor='#EE82EE'><b>class_code</b></td>
                                    <td bgcolor='#EE82EE'><b>transcript_length</b></td>
                                    <td bgcolor='green'><b>Sample_1</b></td>
                                    <td bgcolor='green'><b>Sample_2</b></td>
                                    <td bgcolor='green'><b>Sample_n</b></td>
                                </tr>
                            </thead>
                            <tr>
                                <td>Gene_00002</td>
                                <td>chr1:1431363-1431403</td>
                                <td>NM_031921</td>
                                <td>c</td>
                                <td>&nbsp;</td>
                                <td>161.726</td>
                                <td>20.44</td>
                                <td>81.2435</td>                              
                            </tr>
                            <tr>
                                <td>Gene_00003</td>
                                <td>chr1:2495127-2495222</td>
                                <td>NM_003820</td>
                                <td>c</td>
                                <td>&nbsp;</td>
                                <td>1786.3</td>
                                <td>285.454</td>
                                <td>482.786</td>
                            </tr>
                            <tr>
                                <td>Gene_00004</td>
                                <td>chr1:5446956-5447187</td>
                                <td>-</td>
                                <td>-</td>
                                <td>&nbsp;</td>
                                <td>141.803</td>
                                <td>47.344</td>
                                <td>107.779</td>
                            </tr>
                        </table>
                        <p/>
                    </li>
                    <li>
                        <b>Count</b>
                        <p>
                            Discrete counts of sequence Reads aligned in a genomic region. The program
                            calculates the corrsponding RPKM values for each observation, each sample.
                            A <b>'Library Size file' is optional</b>. If one is not provided, the sum of 
                            Counts in each sample is used as the lbrary size. An example:
                        </p>
                        <table border='1'>
                            <thead>
                                <tr>
                                    <td bgcolor='#C0C0C0'><b>tracking_id</b></td>
                                    <td bgcolor='#C0C0C0'><b>locus</b></td>
                                    <td bgcolor='#C0C0C0'><b>nearest_ref_id</b></td>
                                    <td bgcolor='#EE82EE'><b>class_code</b></td>
                                    <td bgcolor='#EE82EE'><b>transcript_length</b></td>
                                    <td bgcolor='green'><b>Sample_1</b></td>
                                    <td bgcolor='green'><b>Sample_2</b></td>
                                    <td bgcolor='green'><b>Sample_n</b></td>
                                </tr>
                            </thead>
                            <tr>
                                <td>Gene_00002</td>
                                <td>chr1:1431363-1431403</td>
                                <td>NM_031921</td>
                                <td>c</td>
                                <td>&nbsp;</td>
                                <td>20</td>
                                <td>81</td>
                                <td>9</td>                              
                            </tr>
                            <tr>
                                <td>Gene_00003</td>
                                <td>chr1:2495127-2495222</td>
                                <td>NM_003820</td>
                                <td>c</td>
                                <td>&nbsp;</td>
                                <td>285</td>
                                <td>482</td>
                                <td>69</td>
                            </tr>
                            <tr>
                                <td>Gene_00004</td>
                                <td>chr1:5446956-5447187</td>
                                <td>-</td>
                                <td>-</td>
                                <td>&nbsp;</td>
                                <td>47</td>
                                <td>107</td>
                                <td>32</td>
                            </tr>
                        </table>
                        <p/>
                    </li>
                    <li>
                        <b>RPKM &amp; Count</b>
                        <p>
                            Both RPKM and Discrete counts of sequence Reads are provided by the user. This format is useful when
                            the user does nto want to provide those values using custom methods without relying on the system. 
                            There are 2 strict <b>requirements</b>:
                        </p>
                        <ol>
                            <li>The sample names of the 2 columns should be the <b>same</b>.</li>
                            <li>The <b>first</b> coulums is treated as RPKM and the <b>second</b> one as Count.</li>
                        </ol>
                        
                        <p>
                            A <b>'Library Size file' is optional</b>.
                            An example:
                        </p>
                        <table border='1'>
                            <thead>
                                <tr>
                                    <td bgcolor='#C0C0C0'><b>tracking_id</b></td>
                                    <td bgcolor='#C0C0C0'><b>locus</b></td>
                                    <td bgcolor='#C0C0C0'><b>nearest_ref_id</b></td>
                                    <td bgcolor='#EE82EE'><b>class_code</b></td>
                                    <td bgcolor='#EE82EE'><b>transcript_length</b></td>
                                    <td bgcolor='green'><b>Sample_1</b></td>
                                    <td bgcolor='green'><b>Sample_1</b></td>
                                    <td bgcolor='#00FFFF'><b>Sample_2</b></td>
                                    <td bgcolor='#00FFFF'><b>Sample_2</b></td>
                                    <td bgcolor='#FFA500'><b>Sample_n</b></td>
                                    <td bgcolor='#FFA500'><b>Sample_n</b></td>
                                </tr>
                            </thead>
                            <tr>
                                <td>Gene_00002</td>
                                <td>chr1:1431363-1431403</td>
                                <td>NM_031921</td>
                                <td>c</td>
                                <td>&nbsp;</td>
                                <td>2323.12</td>
                                <td>20</td>
                                <td>323.12</td>
                                <td>81</td>
                                <td>223.12</td>
                                <td>9</td>                              
                            </tr>
                            <tr>
                                <td>Gene_00003</td>
                                <td>chr1:2495127-2495222</td>
                                <td>NM_003820</td>
                                <td>c</td>
                                <td>&nbsp;</td>
                                <td>453.12</td>
                                <td>285</td>
                                <td>879.12</td>
                                <td>482</td>
                                <td>223.12</td>
                                <td>69</td>
                            </tr>
                            <tr>
                                <td>Gene_00004</td>
                                <td>chr1:5446956-5447187</td>
                                <td>-</td>
                                <td>-</td>
                                <td>&nbsp;</td>
                                <td>443.12</td>
                                <td>47</td>
                                <td>2323.12</td>
                                <td>107</td>
                                <td>623.12</td>
                                <td>32</td>
                            </tr>
                        </table>
                        <p/>
                    </li>
                    <li>
                        <b>FPKM &amp; Count </b>
                        <p>
                            This uses another commonly used unit of HTS expression data called FPKM (Fragments per kilobase per million).
                            Our program currently does not support an auto conversion from FPKM to Counts. As a result if the user
                            wants to load this kind of data, one has to provide the corresponfing Count value as well.
                            There are 2 strict <b>requirements</b>:
                        </p>
                        <ol>
                            <li>The sample names of the 2 columns should be the <b>same</b>.</li>
                            <li>The <b>first</b> coulums is treated as RPKM and the <b>second</b> one as Count.</li>
                        </ol>
                        
                        <p>
                            A <b>'Library Size file' is optional</b>.
                            An example:
                        </p>
                        <table border='1'>
                            <thead>
                                <tr>
                                    <td bgcolor='#C0C0C0'><b>tracking_id</b></td>
                                    <td bgcolor='#C0C0C0'><b>locus</b></td>
                                    <td bgcolor='#C0C0C0'><b>nearest_ref_id</b></td>
                                    <td bgcolor='#EE82EE'><b>class_code</b></td>
                                    <td bgcolor='#EE82EE'><b>transcript_length</b></td>
                                    <td bgcolor='green'><b>Sample_1</b></td>
                                    <td bgcolor='green'><b>Sample_1</b></td>
                                    <td bgcolor='#00FFFF'><b>Sample_2</b></td>
                                    <td bgcolor='#00FFFF'><b>Sample_2</b></td>
                                    <td bgcolor='#FFA500'><b>Sample_n</b></td>
                                    <td bgcolor='#FFA500'><b>Sample_n</b></td>
                                </tr>
                            </thead>
                            <tr>
                                <td>Gene_00002</td>
                                <td>chr1:1431363-1431403</td>
                                <td>NM_031921</td>
                                <td>c</td>
                                <td>&nbsp;</td>
                                <td>2323.12</td>
                                <td>20</td>
                                <td>323.12</td>
                                <td>81</td>
                                <td>223.12</td>
                                <td>9</td>                              
                            </tr>
                            <tr>
                                <td>Gene_00003</td>
                                <td>chr1:2495127-2495222</td>
                                <td>NM_003820</td>
                                <td>c</td>
                                <td>&nbsp;</td>
                                <td>453.12</td>
                                <td>285</td>
                                <td>879.12</td>
                                <td>482</td>
                                <td>223.12</td>
                                <td>69</td>
                            </tr>
                            <tr>
                                <td>Gene_00004</td>
                                <td>chr1:5446956-5447187</td>
                                <td>-</td>
                                <td>-</td>
                                <td>&nbsp;</td>
                                <td>443.12</td>
                                <td>47</td>
                                <td>2323.12</td>
                                <td>107</td>
                                <td>623.12</td>
                                <td>32</td>
                            </tr>
                        </table>
                        <p/>
                    </li>
                </ol>
            </li>
            <li>
                <b>Species</b>: Currently we porvide annotation support for Human ans Mouse data only. 
                Other requirments of Ref Genome type and build versions apply.
                <ol>
                    <li>Human</li>
                    <li>Mouse</li>
                </ol>
            </li>
            <li>
                <b>Reference Genome</b>:
                Whis refrence genome was used to map align the reads. The same ref genome would be used for other annotaiton.
                We currently support RefSeq and ENSMBL models. This is <b>important as the 'nearest_ref_id' field is used to
                    link to the annotaiton DB and an incorrect selection would lead to undesirable results</b>.
                <ol>
                    <li>RefSeq</li>
                    <li>ENSMBL</li>
                </ol>
            </li>
            <li>
                <b>UCSC Build</b> This field specifies the Reference Genome <b>version</b> used to map/aligh the reads. It is <b></b>important to
                specifiy this correctly</b> as the mapping information changes enough between versions to lead to mapping mistakes.
                <ol>
                    <li>Human: hg19 &amp; hg18</li>
                    <li>Mouse: mm9 &amp; mm8</li>
                </ol>
            </li>
            <li>
                <b>Read Length</b>
                Optional Filed to specify the length of sequence reads in the experiments. For future use.
            </li>
        </ul>
        
        <h4>Library Size File</h4>
        <p>
            This should be a <b>tab-delimited</b> file <b>wihtout</b> header. Each row should have <b>2</b> columns: <b>Sample name</b> and <b>Library Size</b>.
            Comment lines are OK and should start with <em><b>"#"</b></em>. An example:
        </p>
        <table border='1'>
            <tr>
                <td colspan="2"><b># This is a  comment</b></td>
            </tr>
            <tr>
                <td><b>Sample_1</b></td>
                <td>5454545</td>                              
            </tr>
            <tr>
                <td><b>Sample_2</b></td>
                <td>694545</td>
            </tr>
            <tr>
                <td><b>Sample_n</b></td>
                <td>3245443</td>
            </tr>
            <tr>
                <td colspan="2"><b># This is the End</b></td>
            </tr>
        </table>
        <h4>Count to RPKM and vice versa <a name='convert'></a></h4>
        <p>
            When either RPKM or Count information is provided MeV calculates the other based on the publication by
            <b>Mortazavi et al. <i>Nature Methods - 5, 621 - 628 (2008).</i></b> The supplemntary section describes the
            approach in detail. Here is the basic formula used:
            
            <b>RPKM = Count/Library Size/TranscriptLength*1e+9</b>
        </p>
        <b>Rules and requirements</b>
        <ul>
            <li>When RPKM is provided, A library Size file is requred. When Count is provided the file is optional and
                MeV takes the sum of the counts of each sample as the library size.</li>
            <li>When <b>transcript_length</b> annotation column is left empty in the data file MeV calculates the
                same from the <b>locus</b> as the diff between start and end BP.</li>
        </ul>
        <div class="navigation-bar">
			<a href="Loading-agilent-data.html">Previous: Loading Agilent Data</a> | 
			<a class="toc" href="TableofContents.html">Table of Contents</a> | 
			<a class="next" href="Adjustingthedata.html">Next: Adjusting the Data</a>
		</div>
        
    </body>
</html>