<html>
	<head>
		<title>MINET: Mutual Information Network</title>
		<link rel="stylesheet" href="mev manual.css">
		</head>
		</style>
		<body>
<div class="navigation-bar">
	<a class="prev" href="rp.html">Previous: RP-Rank Products</a> | 
	<a class="toc" href="TableofContents.html">Table of Contents</a> | 
	<a class="next" href="edger.html">Next: EdgeR: Empirical Analysis of RNASeq Data in R</a>
</div>

<h3>MINET:Mutual Information Network</h3>
<p>(Patrick E. Meyer, Frederic Lafitte, and Gianluca Bontempi. minet: A R/Bioconductor Package for Inferring Large Transcriptional Networks Using Mutual Information. BMC Bioinformatics, Vol 9, 2008.)</p>
<p>For a given dataset, minet infers the network in two steps. First, the mutual information between all pairs of variables in dataset is computed according to the estimator argument. Then the algorithm given by method considers the estimated mutual information in order to build the network.</p>
	<div class="screenshots ">
		<a href="m1.jpg">
					<img src="m1.jpg">
				</a>
				<br/>
				<span class="caption">MINET Initialization Dialog box</span>
				</div>
				
	<a name="para">
		<h2>Parameters</h2>
	</a>
	<ol>
		<li><b>Dataset</b>: A cluster of genes.</li>
		<li><b>Method</b>: The inference algorithm: 
			<ol>
			<li><b>CLR: </b>
				The CLR algorithm is an extension of relevance network. This algorithm computes the mutual information for each pair 
				</br>of genes and derives a score related to the empirical distribution of the MI values.Instead of considering the mutual 
				</br>information I(Xi;Xj) between features Xi and Xj, it takes into account the score from the mean and the standard deviation 
				</br>of the empirical distribution of the mutual information values I(Xi;Xk), k=1,...,n.
			</li>
			<li><b>Aracne: </b>
				The Aracne procedure starts by assigning to each pair of nodes a weight equal to their mutual
				</br>information. Then, the weakest edge of each triplet is interpreted as an indirect interaction and is
				</br>removed if the difference between the two lowest weights is above a threshold eps.
			</li>
			<li><b>MRNET: </b>
				The MRNET approach consists in repeating a MRMR (Maximum Relevance Minimum Redundancy) feature selection procedure 
				</br>for each variable of the dataset. The MRMR method starts by selecting the variable Xi having the highest mutual
				</br>information with the target Y. In the following steps, given a set S of selected variables, the criterion
				</br>updates S by choosing the variable Xk that maximizes the MRMR score. The weight of each pair Xi;Xj will be the maximum 
				</br>score between the one computed when Xi is the target and the one computed when Xj is the target.
			</li>
			<li><b>MRNETB: </b>
				MRNETB takes the mutual information matrix as input in order to infer the network using the maximum
				</br>relevance/minimum redundancy feature selection method - see MRNET. 
				</br>This method is a variant of MRNET (not yet available).
			</li>
			</ol>
		</li>
		<li><b>Estimator</b>: The name of the entropy estimator to be used for mutual information computation: 
				</br>"mi.empirical", "mi.mm", "mi.shrink", "mi.sg", "spearman", "kendall", "pearson".
		</li> 
		<li><b>Discretization</b>: The name of the discretization method to be used, if required by the estimator 
				</br>"none", "equalfreq", "equalwidth" or "globalequalwidth".
		</li>
		<li><b>Number of Bins</b>:  Integer specifying the number of bins to be used for the discretization 
				</br>if disc is set properly. By default the number of bins is set to sqrt(N) where N is the number of samples.
		</li>
		<li><b>Output</b>: Minet returns a matrix which is the weighted adjacency matrix of the network. 
				</br>The weights range from 0 to 1 and can be seen as a confidence measure on the presence of the arcs. 
				</br>In order to display the network, MeV launches Cytoscape via webstart and displays the network in that application.
		</li>
	</ol>
	</p>
	<h5>Output</h5>
	<p>Minet returns a matrix which is the weighted adjacency matrix of the network. The weights range from 0 to 1 and can be seen as a confidence measure on the presence of the arcs. In order to display the network, MeV launches Cytoscape via webstart and displays the network in that application.</p>
<h5>How to Run MINET</h5>
<ol>
	<div class="screenshots ">
		<a href="minet2.jpg">
					<img src="minet2.jpg">
				</a>
				<br/>
				<span class="caption">Cytoscape</span>
				</div>
<li>Create Cluster</li>
<li>Launch MINET from ToolBar-->Statistics-->MINET</li>
<li>Un-select samples to be excluded from analysis</li>
<li>Select one cluster from Cluster selection Panel</li>
<li>Click on Parameters tab to change method, estimator etc or leave them as is to accept default</li>
<li>Hit OK</li>
</ol>
<h5>Viewing the Network</h5>
	<p>The network is displayed in Cytoscape. IN the initial view all nodes are stacked up on top of each other. In Cytoscape follow <span class="widget2">MenuBar-->Layout-->File-–>Organic</span> (or any layout of your choice) to layout the nodes. 
<div class="navigation-bar">
	<a class="prev" href="rp.html">Previous: RP-Rank Products</a> | 
	<a class="toc" href="TableofContents.html">Table of Contents</a> | 
	<a class="next" href="edger.html">Next: EdgeR: Empirical Analysis of RNASeq Data in R</a>
</div>
</body>
</html>





