<html><body>
<p><font size = "+1"><b>HCL</b></font> -  
<b>Hierarchical Clustering</b> 
<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=9843981&amp;query_hl=4&amp;itool=pubmed_docsum"> 
<br>Eisen, M.B., P.T. Spellman, P.O. Brown, and D. Botstein (1998)  
</a> 
Cluster analysis and display of genome-wide expression patterns.  
<i>Proceedings of the National Academy of Sciences USA </i>  
95:14863-14868. 
<br><a href="http://bioinformatics.oxfordjournals.org/cgi/screenpdf/17/suppl_1/S22"> 
<br>Bar-Joseph, Z., D.K. Gifford, T.S. Jaakkola (2001) 
</a> 
Fast optimal leaf ordering for hierarchical clustering.  
<i>Bioinformatics </i>  
Vol. 17 no. 90001 2001:S22-S29 
        
<br></p><p><font size = "+1"><b>ST</b></font> -  
<b>Support trees (Bootstrapping)</b> 
<br>Graur, D., and W.-H. Li. 2000.  
<i>Fundamentals of Molecular Evolution. Second Edition.</i> 
Sinauer Associates, Sunderland, MA. pp 209-210 
        
<br></p><p><font size = "+1"><b>SOTA</b></font> -  
        
<b>Self-organizing trees</b> 
        
<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=11238068&amp;query_hl=7&amp;itool=pubmed_docsum"> 
<br>Herrero J., A. Valencia, J. Dopazo (2001). 
</a> 
A hierarchical unsupervised growing neural network for clustering gene 
expression patterns.  <i>Bioinformatics</i> 17(2):126-136. 
        
<br></p><p><a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=9069183&amp;query_hl=9&amp;itool=pubmed_docsum"> 
Dopazo, J., and J. M. Carazo (1997). 
</a> 
Phylogenetic reconstruction using an unsupervised growing neural network that adopts the topology  
of a phylogenetic tree.  <i>Journal of Molecular Evolution</i> 44:226-233. 
 
<br></p><p><font size = "+1"><b>KMC</b></font> -  
<b>K-Means</b> 
<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=10783168&amp;query_hl=11&amp;itool=pubmed_docsum"> 
<br>Soukas, A., P. Cohen, N.D. Socci, and J.M. Friedman (2000)  
</a> 
Leptin-specific patterns of gene expression in white adipose tissue.  
<i>Genes and Development</i> 14:963-980. 
        
<br></p><p><font size = "+1"><b>SOM</b></font> -  
<b>Self-Organizing Maps</b> 
<br>Kohonen, T. (1982)   
Self-organized formation of topologically correct feature maps. <i>Biological Cybernetics</i> 43:59-69 
<br></p><p><a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=10077610&amp;query_hl=20&amp;itool=pubmed_docsum"> 
Tamayo, P., D. Slonim, J. Masirov, Q. Zhu, S. Kitareewan,  
</a> 
E. Dmitrovsky, E.S. Lander, and T.R. Golub (1999)   
Interpreting patterns of gene expression with self-organizing maps: Methods and application  
to hematopoietic differentiation. <i>Proceedings of the National Academy of Sciences USA </i>  
96:2907-2912.
        
<br></p><p><font size = "+1"><b>CAST</b></font> - 
<b>Clustering Affinity Search Technique</b>
<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=10582567&amp;query_hl=23&amp;itool=pubmed_docsum"> 
<br>Ben-Dor, A., R. Shamir, and Z. Yakhini (1999) 
</a>
Clustering gene expression patterns. 
<i>Journal of Computational Biology</i> 6:281-297.
        
<br></p><p><font size = "+1"><b>QTC</b></font> - 
<b>Jackknife Clustering</b>
<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=10568750&amp;query_hl=26&amp;itool=pubmed_docsum" 
>
<br>Heyer, L.J., S. Kruglyak, and S. Yooseph (1999) 
</a>
Exploring expression data: identification and analysis of coexpressed genes. 
<i>Genome Research</i> 9:1106-1115.
        
<br></p><p><font size = "+1"><b>GSH</b></font> - 
<b>Gene shaving</b>
<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=11178228&amp;query_hl=28&amp;itool=pubmed_docsum"> 
<br>Hastie, T. et al. (2000).
</a>
'Gene shaving' as a method for identifying distinct sets of genes with similar expression patterns 
Genome Biology 1(2):research0003.1-0003.21. 
        
<br></p><p><font size = "+1"><b>FOM</b></font> -  
<b>Figures of Merit</b> 
<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=11301299&amp;query_hl=31&amp;itool=pubmed_docsum"> 
<br>Yeung, K.Y., D.R. Haynor, and W.L. Ruzzo (2001)  
</a> 
Validating clustering for gene expression data.  
<i>Bioinformatics</i> 17:309-318. 
        
<br></p><p><font size = "+1"><b>PCA</b></font> -  
<b>Principal Components Analysis</b> 
<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=10902193&amp;query_hl=33&amp;itool=pubmed_docsum" 
> 
<br>Raychaudhuri, S.,  J. M. Stuart, &amp; R. B. Altman (2000)  
</a> 
Principal components analysis to summarize microarray experiments:  
application to sporulation time series.  
<i>Pacific Symposium on Biocomputing 2000, Honolulu, Hawaii,</i> 452-463.  
Available at http://smi-web.stanford.edu/pubs/SMI_Abstracts/SMI-1999-0804.html 
        
<br></p><p><font size = "+1"><b>COA</b></font> -  
<b>Correspondence Analysis</b> 
<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=11535808&amp;query_hl=35&amp;itool=pubmed_docsum"> 
<br>Fellenberg K. et al. (2001)  
</a> 
Correspondence analysis applied to microarray data.  
<i>Proceedings of the National Academy of Sciences USA</i> 98: 10781-10786 
<br></p><p><a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=12490444&amp;query_hl=37&amp;itool=pubmed_docsum"> 
Culhane A.C. et al. (2002)  
</a> 
Between-group analysis of microarray data.  
<i>Bioinformatics</i> 18:1600-1608. 
        
<br></p><p><font size = "+1"><b>RN</b></font> -  
<b>Relevance networks</b> 
<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=11027309&amp;query_hl=39&amp;itool=pubmed_docsum" 
> 
<br>Butte, A.J., P. Tamayo, D. Slonim, T.R. Golub and I.S. Kohane (2000)  
</a> 
Discovering functional relationships between RNA expression and chemotherapautic  
susceptibility using relevance networks. 
<i>Proceedings of the National Academy of Sciences USA</i> 97: 12182-12186 
        
<br></p><p><font size = "+1"><b>PTM</b></font> -  
<b>Template Matching</b> 
<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=11597334&amp;query_hl=42&amp;itool=pubmed_docsum" 
> 
<br>Pavlidis, P., and W.S. Noble (2001)  
</a> 
Analysis of strain and regional variation in gene expression in mouse brain.  
<i>Genome Biology</i> 2:research0042.1-0042.15 
        
<br></p><p><font size = "+1"><b>SAM</b></font> -  
<b>Significance Analysis of Microarrays</b> 
<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=11309499&amp;query_hl=44&amp;itool=pubmed_docsum" 
> 
<br>Tusher, V.G., R. Tibshirani and G. Chu. (2001)  
</a> 
Significance analysis of microarrays applied to the ionizing radiation response.  
<i>Proceedings of the National Academy of Sciences USA</i> 98: 5116-5121. 
        
<br></p><p><a href="http://www-stat.stanford.edu/~tibs/SAM/">
Chu, G., B. Narasimhan, R. Tibshirani and V. Tusher (2002).
</a> 
SAM "Significance Analysis of Microarrays" Users Guide and Technical Document. 
http://www-stat.stanford.edu/~tibs/SAM/ 

<br></p><p><font size = "+1"><b>ANOVA</b></font> -  
<b>One-way Analysis of Variance</b>   
        
<br>Zar, J.H. 1999. Biostatistical Analysis. 4th ed. Prentice Hall, NJ. 
        
<br></p><p><font size = "+1"><b>TTEST</b></font> -  
<b>T-Tests</b> 
<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=12016052&amp;query_hl=48&amp;itool=pubmed_docsum" 
> 
<br>Pan, W. (2002).   
</a> 
A comparative review of statistical methods for discovering differentially expressed genes in replicated microarray experiments  
<i>Bioinformatics</i> 18: 546-554. 
        
<br></p><p>Dudoit, S., Y.H. Yang, M.J. Callow, and T. Speed (2000). 
Statistical methods for identifying differentially expressed genes  
in replicated cDNA microarray experiments. <i>Technical report 2000</i>  
Statistics Department, University of California, Berkeley. 

<br></p><p>Welch, B.L. (1947). 
The generalization of ‘students’ problem when several different population variances are involved.  
<i>Biometrika</i> 34: 28-35. 
<br></p><p> <a href="http://linus.nci.nih.gov/~brb/TechReport.htm" 
> 
Korn, E.L., J.F. Troendle, L.M. McShane, R. Simon (2001). 
</a> 
Controlling the number of false discoveries: application to high-dimensional genomic data  
<i> Technical report 003</i>, Biometric Research Branch, National Cancer Institute. http://linus.nci.nih.gov/~brb/TechReport.htm            
        
<br></p><p>Korn, E.L., J.F. Troendle, L.M. McShane, R. Simon (2004). 
Controlling the number of false discoveries: application to high-dimensional genomic data  
<i>Journal of Statistical Planning and Inference</i> 124: 379-398.       

<br></p><p><font size = "+1"><b>TFA</b></font> -  
<b>Two Factor ANOVA</b> 
        
<br>Keppel, G., and S. Zedeck. (1989). Data Analysis for Research Designs. W. H. Freeman and Co., NY.                
<br></p><p>Manly, B.F.J. (1997). Randomization, Bootstrap and Monte Carlo Methods in Biology. 2nd ed. Chapman and Hall / CRC , FL.        
<br></p><p>Zar, J.H. 1999. Biostatistical Analysis. 4th ed. Prentice Hall, NJ.   
                
<br></p><p><font size = "+1"><b>SVM</b></font> -  
<b>Support Vector Machines</b> 
<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=10618406&amp;query_hl=86&amp;itool=pubmed_docsum"> 
<br>Brown, M.P., W.N. Grundy, D. Lin, N. Cristianini, C.W. Sugnet,  
T.S. Furey, M. Ares, Jr., and D. Haussler (2000)  
</a> 
Knowledge-based analysis of microarray gene expression data by using support vector machines.  
<i>Proceedings of the National Academy of Sciences USA</i> 97: 262-267. 
        
<br></p><p><font size = "+1"><b>KNNC</b></font> -  
<b>K-Nearest Neighbor Classification</b> 
<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=11779842&amp;query_hl=75&amp;itool=pubmed_docsum" 
> 
<br>Theilhaber, J., Connolly, S. Roman-Roman, S. Bushnell, A. Jackson, K. Call, T. Garcia, R. Baron (2002)    
</a> 
Finding Genes in the C2Cl2 Osteogenic Pathway by K-Nearest-Neighbor Classification of Expression Data. 
<i>Genome Research</i> 12:165-176 

<br></p><p><font size = "+1"><b>DAM</b></font> -  
<b>Discriminant Analysis Module</b> 
<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=12217913&amp;query_hl=78&amp;itool=pubmed_docsum"> 
<br>Nguyen, D.V., D.M. Rocke.  
</a> 
Multi-class Cancer Classification via Partial Least Squares with Gene Expression Profiles. 
<i>Bioinformatics</i> 18(9):1216-1226, 2002 
       
<br></p><p><font size = "+1"><b>TRN</b></font> -  
<b>Expression Terrain Maps</b> 
<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=11557892&amp;query_hl=82&amp;itool=pubmed_docsum"> 
<br>Kim, S.K., J. Lund, M. Kiraly, K. Duke, M. Jiang, J.M. Stuart, A. Eizinger, 
B.N. Wylie, and G.S. Davidson (2001)  
</a> 
A Gene Expression Map for Caenorhabditis elegans.  
<i>Science</i> 293: 2087-2092. 

<br></p><p><font size = "+1"><b>EASE</b></font> -  
<b>Expression Analysis Systematic Explorer</b> 
<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=14519205&amp;query_hl=84&amp;itool=pubmed_docsum"> 
<br>Hosack, D.A., G. Dennis Jr., B.T. Sherman, H.C. Lane, R.A. Lempicki. (2001) 
</a> 
Identifying biological themes within lists of genes with EASE. 
<i>Genome Biol.</i> 4:R70-R70.8, 2003.

<br></p><p><font size = "+1"><b>nEASE</b></font> -  
<b>An addition to the EASE module</b> 
<a href="">
<br>Chittenden, T.W., E.A. Howe, J. Taylor, J. Mar, M. Aryee, J. Braisted, S. Nair, J. Quackenbush, C. Holmes. (2009) 
</a> 
Molecular Subtype-specific Gene Ontology Classification in Human Breast Cancer 
<i>Submitted.</i> 

<p><font size = "+1"><b>TEASE</b></font> -
<b>Tree EASE</b></p><p>
Gottardo R, Raftery A.E, Yeung K.Y, and R.E. Bumgarner, Quality Control and Robust Estimation for cDNA Microarrays with Replicates, accepted for publication in The Journal of the American Statistical Association
<br></p><p>
<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=16542223&amp;query_hl=84&amp;itool=pubmed_docsum">
Gottardo R, Raftery AE, Yee Yeung K, Bumgarner RE. 2006. Bayesian robust inference for differential gene expression in microarrays with multiple samples.</a>
Biometrics. 62(1):10-8.

<br></p><p><font size = "+1"><b>USC</b></font> -
<b>Uncorrelated Shrunken Centroid</b></p><p>
Yeung KY, Bumgarner RE.  
<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=14659020&amp;query_hl=84&amp;itool=pubmed_docsum">Multiclass classification of microarray data with repeated measurements: application to cancer.</a>
 Genome Biol. 2003;4(12):R83. Epub 2003 Nov 24.
<br></p><p>
Yeung KY, Bumgarner RE. 
<a href="http://ukpmc.ac.uk/articlerender.cgi?artid=534920">Correction: Multiclass classification of microarray data with repeated measurements: application to cancer.</a>
 Genome Biol. 2005;6(13):405. Epub 2006 Jan 3.

</p><p><font size = "+1"><b>CGH</b></font> -
<b>Comparitive Genomic Hybridization</b></p><p>
Margolin AA, Greshock J, Naylor TL, Mosse Y, Maris JM, Bignell G, Saeed AI, Quackenbush J, Weber BL. 
<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=15905276&amp;query_hl=84&amp;itool=pubmed_docsum">CGHAnalyzer: a stand-alone software package for cancer genome analysis using array-based DNA copy number data.</a> 
<i>Bioinformatics.</i> 2005 Aug 1;21(15):3308-11.
</p>

<p><font size = "+1"><b>NonpaR</b></font></p><p>
Hollander M, Wolfe D. 1999 Nonparametric Statistical Methods. 1999.
</p>

<p><font size = "+1"><b>BN/LM</b></font> -
<b>Bayes Networks / Literature Mining</b></p><p>
Djebbari, A., Quackenbush, J. In Review. Seeded Bayesian Networks: constructing genetic networks from microarray data.
</p>

<p><font size = "+1"><b>BETR</b></font> -
<b>Bayesian Estimation of Temporal Regulation</b></p><p>
Aryee, M., J. Gutierrez-Pabello, I. Kramnik, T., Maiti, J. Quackenbush 2008. 
An Improved Empirical Bayes Approach to Estimating Differential Expression in Microarray Time-Course Data:  
BETR (Bayesian Estimation of Temporal Regulation). BMC Bioinformatics,  accepted Oct. 2009.
</p>

<p><font size = "+1"><b>RP</b></font> -
<b>Rank Products</b></p><p>
Breitling, R., P. Armengaud, A. Amtmann, P. Herzyk 2004. Rank products: a simple, yet powerful, new method to detect differentially regulated genes in replicated microarray experiments.  Federation of European Biochemical Societies. 83-92.
</p>

<p><font size = "+1"><b>GSEA</b></font> -
<b>Gene Set Enrichment Analysis</b></p><p>
Jiang, Z. and Gentleman, R. (2007). Extensions to gene set enrichment analysis. Bioinformatics, 23, 306–313.
</p>

<p><font size = "+1"><b>R-CRAN</b></font> -
<b>R and BioConductor</b></p><p>
R Development Core Team (2009). R: A language and environment for statistical computing. R Foundation for Statistical Computing, Vienna, Austria. ISBN 3-900051-07-0, URL <a href="http://www.R-project.org">http://www.R-project.org</a>.
</p>

<p><font size = "+1"><b>LIMMA</b></font> -
<b>LIMMA algorithm</b></p><p>
Smyth, G. K. (2004). Linear models and empirical Bayes methods for assessing differential expression in microarray experiments. Statistical Applications in Genetics and Molecular Biology 3, No. 1, Article 3.
</p>

<p><font size = "+1"><b>LIMMA</b></font> -
<b>LIMMA R software</b></p><p>
Smyth, G. K. (2005). Limma: linear models for microarray data. In:  Bioinformatics and Computational Biology Solutions using R and Bioconductor, R. Gentleman, V. Carey, S. Dudoit, R. Irizarry, W. Huber (eds.), Springer, New York, pages 397-420.
</p>

<p><font size = "+1"><b>NMF</b></font> -
<b>Non-negative Matrix Factorization</b></p><p>Brunet, J-P., Tamayo, P., Golub, T.R., and Mesirov, J.P. 2004. 
Metagenes and molecular pattern discovery using matrix factorization. <i>Proc. Natl. Acad. Sci. USA</i> 101(12):4164Â–4169. 
<br><br>
Devarajan K, 2008 Nonnegative Matrix Factorization: An Analytical and Interpretive Tool in Computational Biology. PLoS Comput Biol 4(7): e1000029. doi:10.1371/journal.pcbi.1000029 
<br><br>
Lee DD, Seung SH (2001) Algorithms for nonnegative matrix factorization. Adv Neural Inform Process Syst 13: 556–562.

</p>
</body></html>